fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/impl/lookup/entity/EntityApiLookupImpl.java

```diff
@@ -0,0 +1,162 @@
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.lookup.v1.custom.ApiLookupMap;
+import net.fabricmc.fabric.api.lookup.v1.custom.ApiProviderMap;
+import net.fabricmc.fabric.api.lookup.v1.entity.EntityApiLookup;
+
+public class EntityApiLookupImpl<A, C> implements EntityApiLookup<A, C> {
+	private static final Logger LOGGER = LogManager.getLogger("fabric-api-lookup-api-v1/entity");
+	private static final ApiLookupMap<EntityApiLookup<?, ?>> LOOKUPS = ApiLookupMap.create(EntityApiLookupImpl::new);
+	private static final Map<Class<?>, Set<EntityType<?>>> REGISTERED_SELVES = new HashMap<>();
+	private static boolean checkEntityLookup = true;
+
+	private final Class<A> apiClass;
+	private final Class<C> contextClass;
+	private final ApiProviderMap<EntityType<?>, EntityApiProvider<A, C>> providerMap = ApiProviderMap.create();
+	private final List<EntityApiProvider<A, C>> fallbackProviders = new CopyOnWriteArrayList<>();
+
+	private EntityApiLookupImpl(Class<A> apiClass, Class<C> contextClass) {
+		this.apiClass = apiClass;
+		this.contextClass = contextClass;
+	}
+
+	@SuppressWarnings("unchecked")
+	public static <A, C> EntityApiLookup<A, C> get(Identifier lookupId, Class<A> apiClass, Class<C> contextClass) {
+		return (EntityApiLookup<A, C>) LOOKUPS.getLookup(lookupId, apiClass, contextClass);
+	}
+
+	public static void checkSelfImplementingTypes(MinecraftServer server) {
+		if (checkEntityLookup) {
+			checkEntityLookup = false;
+			REGISTERED_SELVES.forEach((apiClass, entityTypes) -> {
+				for (EntityType<?> entityType : entityTypes) {
+					Entity entity = entityType.create(server.getOverworld());
+
+					if (entity == null) {
+						String errorMessage = String.format(
+								"Failed to register self-implementing entities for API class %s. Can not create entity of type %s.",
+								apiClass.getCanonicalName(),
+								Registry.ENTITY_TYPE.getId(entityType)
+						);
+						throw new NullPointerException(errorMessage);
+					}
+
+					if (!apiClass.isInstance(entity)) {
+						String errorMessage = String.format(
+								"Failed to register self-implementing entities. API class %s is not assignable from entity class %s.",
+								apiClass.getCanonicalName(),
+								entity.getClass().getCanonicalName()
+						);
+						throw new IllegalArgumentException(errorMessage);
+					}
+				}
+			});
+		}
+	}
+
+	@Override
+	@Nullable
+	public A find(Entity entity, C context) {
```

Comment:
Should this check for `EntityPredicates.VALID_ENTITY.matches(entity)` (i.e. `entity.isAlive()`) ? Not sure, but I noticed that hoppers perform this check. I'd say it's better to add it now and then remove it later if it causes issues.